File: electron/sqlite/queries/projects.js
/**
 * 项目 SQLite 查询模块
 *
 * 实现 projects 表的 CRUD 操作。
 * 特点：删除时级联删除 9 张关联表的数据。
 */
import log from 'electron-log';
import { tryGetSqlite, rowToCamel, toSqliteValue } from './helpers';
// ═══════════════════════════════════════════════════════════════════════════════
// 列映射
// ═══════════════════════════════════════════════════════════════════════════════
var PROJ_COLUMNS = {
    id: 'id',
    name: 'name',
    description: 'description',
    address: 'address',
    startDate: 'start_date',
    endDate: 'end_date',
    status: 'status',
    budget: 'budget',
    projectManagerId: 'project_manager_id',
    createdAt: 'created_at',
    updatedAt: 'updated_at',
};
var PROJ_INSERT_COLS = Object.values(PROJ_COLUMNS).filter(function (c) { return c !== 'id'; });
var PROJ_INSERT_PLACEHOLDERS = PROJ_INSERT_COLS.map(function () { return '?'; }).join(', ');
var PROJ_INSERT_SQL = "INSERT INTO projects (".concat(PROJ_INSERT_COLS.map(function (c) { return "\"".concat(c, "\""); }).join(', '), ") VALUES (").concat(PROJ_INSERT_PLACEHOLDERS, ")");
/** 级联删除涉及的表名+项目ID列名 */
var CASCADE_DELETE_TABLES = [
    { table: 'cost_ledger', column: 'project_id' },
    { table: 'settlements', column: 'project_id' },
    { table: 'invoices', column: 'project_id' },
    { table: 'income_contracts', column: 'project_id' },
    { table: 'expense_contracts', column: 'project_id' },
    { table: 'agreement_contracts', column: 'project_id' },
    { table: 'wages', column: 'project_id' },
    { table: 'attendances', column: 'project_id' },
    { table: 'project_members', column: 'project_id' },
];
// ═══════════════════════════════════════════════════════════════════════════════
// 读操作
// ═══════════════════════════════════════════════════════════════════════════════
/** 列出所有项目（含项目经理姓名 JOIN members） */
export function listProjects() {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return null;
    try {
        var rows = sqlite.prepare("\n      SELECT p.*, m.name as project_manager_name\n      FROM projects p\n      LEFT JOIN members m ON p.project_manager_id = m.id\n      ORDER BY p.created_at DESC\n    ").all();
        return rows.map(function (row) {
            var camel = rowToCamel(row);
            // project_manager_name 来自 JOIN，直接挂上
            camel.projectManagerName = row.project_manager_name || '';
            return camel;
        });
    }
    catch (err) {
        log.error('[SQLite] projects.list error:', err);
        return null;
    }
}
// ═══════════════════════════════════════════════════════════════════════════════
// 写操作
// ═══════════════════════════════════════════════════════════════════════════════
/** 创建项目 */
export function createProject(project) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var params = PROJ_INSERT_COLS.map(function (col) {
            var _a;
            var jsonKey = (_a = Object.entries(PROJ_COLUMNS).find(function (_a) {
                var c = _a[1];
                return c === col;
            })) === null || _a === void 0 ? void 0 : _a[0];
            if (!jsonKey)
                return null;
            return toSqliteValue(project[jsonKey]);
        });
        (_a = sqlite.prepare(PROJ_INSERT_SQL)).run.apply(_a, params);
        return true;
    }
    catch (err) {
        log.error('[SQLite] projects.create error:', err);
        return false;
    }
}
/** 更新项目 */
export function updateProject(project) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var setClauses = [];
        var params = [];
        for (var _i = 0, _b = Object.entries(project); _i < _b.length; _i++) {
            var _c = _b[_i], jsonKey = _c[0], value = _c[1];
            if (jsonKey === 'id')
                continue;
            var col = PROJ_COLUMNS[jsonKey];
            if (!col)
                continue;
            setClauses.push("\"".concat(col, "\" = ?"));
            params.push(toSqliteValue(value));
        }
        // 始终更新 updatedAt
        setClauses.push('"updated_at" = ?');
        params.push(new Date().toISOString());
        if (setClauses.length === 1)
            return true; // 仅 updatedAt
        params.push(project.id);
        var sql = "UPDATE projects SET ".concat(setClauses.join(', '), " WHERE id = ?");
        var result = (_a = sqlite.prepare(sql)).run.apply(_a, params);
        return result.changes > 0;
    }
    catch (err) {
        log.error('[SQLite] projects.update error:', err);
        return false;
    }
}
/** 删除项目（含级联删除 9 张关联表，事务） */
export function deleteProject(projectId) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var doDelete = sqlite.transaction(function () {
            // 级联删除关联表数据
            for (var _i = 0, CASCADE_DELETE_TABLES_1 = CASCADE_DELETE_TABLES; _i < CASCADE_DELETE_TABLES_1.length; _i++) {
                var _a = CASCADE_DELETE_TABLES_1[_i], table = _a.table, column = _a.column;
                sqlite.prepare("DELETE FROM \"".concat(table, "\" WHERE \"").concat(column, "\" = ?")).run(projectId);
            }
            // 删除项目本身
            sqlite.prepare('DELETE FROM projects WHERE id = ?').run(projectId);
        });
        doDelete();
        return true;
    }
    catch (err) {
        log.error('[SQLite] projects.delete error:', err);
        return false;
    }
}

File: src-tauri/src/commands/projects.rs
/**
 * 项目管理命令
 *
 * 对应 Electron 版本的 projects.ts
 */

use crate::db::AppState;
use crate::error::{AppError, AppResult};
use rusqlite::params;
use serde::{Deserialize, Serialize};
use tauri::{command, State};

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Project {
    pub id: i64,
    pub name: String,
    pub description: Option<String>,
    pub address: Option<String>,
    pub start_date: Option<String>,
    pub end_date: Option<String>,
    pub status: Option<String>,
    pub budget: f64,
    pub project_manager_id: Option<i64>,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
    pub project_manager_name: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateProject {
    pub name: String,
    pub description: Option<String>,
    pub address: Option<String>,
    pub start_date: Option<String>,
    pub end_date: Option<String>,
    pub status: Option<String>,
    pub budget: Option<f64>,
    pub project_manager_id: Option<i64>,
}

/// 获取所有项目
#[command]
pub fn get_projects(state: State<'_, AppState>) -> AppResult<Vec<Project>> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let mut stmt = db
        .prepare(
            "SELECT p.id, p.name, p.description, p.address, p.start_date, p.end_date,
                    p.status, p.budget, p.project_manager_id, p.created_at, p.updated_at,
                    m.name as project_manager_name
             FROM projects p
             LEFT JOIN members m ON p.project_manager_id = m.id
             ORDER BY p.id DESC",
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    let projects = stmt
        .query_map([], |row| {
            Ok(Project {
                id: row.get(0)?,
                name: row.get(1)?,
                description: row.get(2)?,
                address: row.get(3)?,
                start_date: row.get(4)?,
                end_date: row.get(5)?,
                status: row.get(6)?,
                budget: row.get(7)?,
                project_manager_id: row.get(8)?,
                created_at: row.get(9)?,
                updated_at: row.get(10)?,
                project_manager_name: row.get(11)?,
            })
        })
        .map_err(|e| AppError::Database(e.to_string()))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Database(e.to_string()))?;

    Ok(projects)
}

/// 获取单个项目
#[command]
pub fn get_project(state: State<'_, AppState>, id: i64) -> AppResult<Project> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    db.query_row(
        "SELECT p.id, p.name, p.description, p.address, p.start_date, p.end_date,
                p.status, p.budget, p.project_manager_id, p.created_at, p.updated_at,
                m.name as project_manager_name
         FROM projects p
         LEFT JOIN members m ON p.project_manager_id = m.id
         WHERE p.id = ?1",
        params![id],
        |row| {
            Ok(Project {
                id: row.get(0)?,
                name: row.get(1)?,
                description: row.get(2)?,
                address: row.get(3)?,
                start_date: row.get(4)?,
                end_date: row.get(5)?,
                status: row.get(6)?,
                budget: row.get(7)?,
                project_manager_id: row.get(8)?,
                created_at: row.get(9)?,
                updated_at: row.get(10)?,
                project_manager_name: row.get(11)?,
            })
        },
    )
    .map_err(|e| match e {
        rusqlite::Error::QueryReturnedNoRows => {
            AppError::NotFound(format!("项目 {} 不存在", id))
        }
        _ => AppError::Database(e.to_string()),
    })
}

/// 创建项目
#[command]
pub fn create_project(state: State<'_, AppState>, project: CreateProject) -> AppResult<i64> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;
    let now = chrono::Local::now().format("%Y-%m-%d %H:%M:%S").to_string();

    db.execute(
        "INSERT INTO projects (name, description, address, start_date, end_date, status, budget, project_manager_id, created_at, updated_at)
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10)",
        params![
            project.name,
            project.description,
            project.address,
            project.start_date,
            project.end_date,
            project.status.unwrap_or_else(|| "planning".to_string()),
            project.budget.unwrap_or(0.0),
            project.project_manager_id,
            now,
            now,
        ],
    )
    .map_err(|e| AppError::Database(e.to_string()))?;

    Ok(db.last_insert_rowid())
}

/// 更新项目
#[command]
pub fn update_project(state: State<'_, AppState>, project: Project) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute(
            "UPDATE projects SET name = ?1, description = ?2, address = ?3, start_date = ?4,
             end_date = ?5, status = ?6, budget = ?7, project_manager_id = ?8, updated_at = datetime('now')
             WHERE id = ?9",
            params![
                project.name,
                project.description,
                project.address,
                project.start_date,
                project.end_date,
                project.status,
                project.budget,
                project.project_manager_id,
                project.id,
            ],
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("项目 {} 不存在", project.id)));
    }

    Ok(())
}

/// 删除项目
#[command]
pub fn delete_project(state: State<'_, AppState>, id: i64) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    // 检查是否有关联数据
    let checks = vec![
        ("project_workers", "工人"),
        ("income_contracts", "收入合同"),
        ("expense_contracts", "支出合同"),
        ("invoices", "发票"),
        ("settlements", "结算"),
    ];

    for (table, label) in checks {
        let count: i64 = db
            .query_row(
                &format!("SELECT COUNT(*) FROM {} WHERE project_id = ?1", table),
                params![id],
                |row| row.get(0),
            )
            .unwrap_or(0);

        if count > 0 {
            return Err(AppError::Validation(format!(
                "项目下还有 {} 条{}记录，请先删除",
                count, label
            )));
        }
    }

    let affected = db
        .execute("DELETE FROM projects WHERE id = ?1", params![id])
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("项目 {} 不存在", id)));
    }

    // 级联删除成本台账
    db.execute(
        "DELETE FROM cost_ledger WHERE project_id = ?1",
        params![id],
    )
    .map_err(|e| AppError::Database(e.to_string()))?;

    Ok(())
}
